pkg/kratos/service.go

```diff
@@ -387,6 +393,78 @@ func (s *Service) UpdateRecoveryFlow(
 	return returnToResp, resp.Cookies(), nil
 }
 
+func (s *Service) executeIdentifierFirstUpdateLoginRequest(
```

Comment:
> bypass kratos SDK for updating login and make a low-level http call to kratos with Content-Type: application/x-www-form-urlencoded to avoid hitting [this](https://github.com/ory/kratos/blob/15f3eb3aa25c92e3d1e025082226f26d2ab58718/selfservice/strategy/idfirst/strategy_login.go#L150C2-L154C3) error

question: looking at that sample of code and [this](https://github.com/ory/herodot/blob/master/json.go#L100) function, it looks like even though the response is an error the flow is in there. Can't we use the sdk and get the flow from the response body?

Reply:
Sure:
1. Clone `IAM-1639-identifier-first-pattern` branch and apply [this](https://gist.github.com/natalian98/c35da57481427b652f866094802795d6) patch for frontend on top.
2. In `kratos.yaml`, switch the login style to `identifier_first`.
3. Build and run login ui and the dev docker compose as usual.
4. Go to login page, it will be divided into 2 steps: in first you provide email, next you'll be shown available login methods.

If you want to test calling sdk, just switch line 68 in the patch file to `/api/kratos/self-service/login?flow=${encodeURIComponent(flowId)}`

Reply:
As discussed privately it appears to be a debugger issue as I can also get the flow from logs.
That said, 400 means form validation errors, causing the login flow update to fail. The returned flow object reflects this failure, so I still believe using the SDK isn’t feasible for this method. In a happy path, we should get a 303 and redirect to the next step, just like with my workaround.

Reply:
400 means that the server considers that there was a client error, in this case this means that a wrong content type was requested, not sure if this is a form validation error. IMO you should not get a 303 in a backchannel request, as a backend would not want to be served an html page. When I manually navigate my browser to `<login_ui_url>/ui/login?flow=<flow_id>` (after the sdk call), I can see that the flow has been updated.

I can see the argument that we should not expect an error response in a non-error path, but I think we should consider it since we wouldn't have to implement our own http logic.

Reply:
I posted a question on the ory slack about the 400, let's wait for their reply and discuss next week @shipperizer @BarcoMasile

Reply:
my 2 cents here: relying on a 400 error to handle a successful case is too fragile an approach and error prone. Also, we cannot expect Ory to keep a "successful" behavior when an error code is returned, since we would not have any guarantee that that behavior is kept in the future.
If I understand this thread correctly, then I suggest we don't rely on an error for our case.

Reply:
Great and we already have a function for parsing the flow in case of error. Why don't we parse the flow in order to decide whether an error actually occurred or not?

Reply:
as mentioned in https://github.com/canonical/identity-platform-login-ui/pull/697#discussion_r2272479968 let's tackle this in a separate issue aimed at refactoring the interaction between kratos, backend and frontend.

let's move on from this

cc @nsklikas @natalian98